### ISSUE TITLE
MySQL 8.0: skeema init fails with "directory already exists"

### ISSUE THREAD

**Issue by FlaviuRadulescu** [skeema/skeema#35]
I am trying to use Skeema for the first time and after a  /root/go/bin/skeema init production --debug -h hostname -u root -p I get an error message saying that unable to use directory <directory path>  for schema: Directory <directory path> already exists.

It doesn't matter if delete the folder and try again I get the same message. The folder is created by the above script, and I have a .skeema file in that directory.

One of the problems may be the fact that I am using Mysql 8.0 but still I don't see the connection with the folder name.

**Comment by evanelias** [comment]
Thank you for trying out Skeema and for reporting an issue! Although Skeema [doesn't support MySQL 8.0 yet](https://github.com/skeema/skeema/blob/master/doc/requirements.md#unsupported-features), I would expect a different error than the one you are seeing. Here are a few things to gather more info and debug:

* Can you please try the following and paste the output:
  * `mkdir schemas; cd schemas; skeema init production --debug -h hostname -u root -p; ls -laR; cat hostname/.skeema` (replace hostname as needed)
* When running `skeema init`, if you supply `--dir some-new-dir-name` is the result any different? (this option specifies the name to use for the host-level dir; if you omit it, the hostname will be used instead, and maybe that's conflicting with something)
* Do you actually have multiple schemas (databases) in MySQL or do you just have one? If you only have one, you could try supplying `--schema schemaname` (replace with real schema name) to `skeema init`, which should combine the schema and host level dirs into a single dir. This probably won't fix your problem but it might provide us more insight on what broke, since it sounds like you're getting the error creating the schema-level directory.
* What is the output of `skeema --version`? If you built from source, how recently did you clone? Is the result the same from a [pre-built release](https://github.com/skeema/skeema/releases)?

Thanks!

For background, Skeema doesn't support MySQL 8.0 yet since that requires upgrading the driver library, which then necessitates some other changes relating to the driver's connection retry behavior. I also still need to delve into 8.0's changes, especially around authentication methods, the data dictionary, and a few other new things. I was originally planning on waiting another month or two before delving into MySQL 8.0 support, but I can definitely re-prioritize if there's sufficient user interest!

**Comment by FlaviuRadulescu** [comment]
Dear Evan, thank you for your super fast answer.

Please find bellow answers to your questions:

```
# mkdir schemas; cd schemas; /root/go/bin/skeema init production --debug -h localhost -u root -p;
Enter password:
2018-08-06 21:09:30 [ERROR] Unable to connect to localhost:/tmp/mysql.sock for /var/www/html/folder-name/development-www/schemas/localhost: dial unix /tmp/mysql.sock: connect: no such file or directory
2018-08-06 21:09:30 [DEBUG] Exit code 2
21:09:30 root  /var/www/html/_/development-www/schemas master
# ls -laR; cat localhost/.skeema
.:
total 12
drwxr-xr-x.  3 root root 4096 Aug  6 21:09 .
drwxr-xr-x. 24 root root 4096 Aug  6 21:09 ..
drwxr-xr-x.  2 root root 4096 Aug  6 21:09 localhost

./localhost:
total 8
drwxr-xr-x. 2 root root 4096 Aug  6 21:09 .
drwxr-xr-x. 3 root root 4096 Aug  6 21:09 ..
cat: localhost/.skeema: No such file or directory
```

After this I have added the -S /var/lib/mysql/mysql.sock 

```
#  mkdir schemas; cd schemas; /root/go/bin/skeema init production --debug -S /var/lib/mysql/mysql.sock -h localhost -u root -p;
Enter password:
2018-08-06 21:12:08 [INFO]  Creating and using host dir /var/www/html/folder-name/development-www/schemas/localhost for localhost:/var/lib/mysql/mysql.sock

2018-08-06 21:12:08 [ERROR] Unable to use directory /var/www/html/folder-name/development-www/schemas/localhost for schema : Directory /var/www/html/folder-name/development-www/schemas/localhost already exists
2018-08-06 21:12:08 [DEBUG] Exit code 73


# ls -laR; cat localhost/.skeema
.:
total 12
drwxr-xr-x.  3 root root 4096 Aug  6 21:12 .
drwxr-xr-x. 24 root root 4096 Aug  6 21:12 ..
drwxr-xr-x.  2 root root 4096 Aug  6 21:12 localhost

./localhost:
total 12
drwxr-xr-x. 2 root root 4096 Aug  6 21:12 .
drwxr-xr-x. 3 root root 4096 Aug  6 21:12 ..
-rw-r--r--. 1 root root   88 Aug  6 21:12 .skeema
[production]
socket=/var/lib/mysql/mysql.sock
user=root
flavor=mysql:8.0
host=localhost

```

As you can see the .skeema file is created despite the error message

I have tried with --dir new folder option and I get almost the same error. the .skeema is also created

```# mkdir schemas; cd schemas; /root/go/bin/skeema init production --debug --dir newfolder -S /var/lib/mysql/mysql.sock -h localhost -u root -p
Enter password:
2018-08-06 21:17:28 [INFO]  Creating and using host dir /var/www/html/folder-name/development-www/schemas/newfolder for localhost:/var/lib/mysql/mysql.sock

2018-08-06 21:17:28 [ERROR] Unable to use directory /var/www/html/folder-name/development-www/schemas/newfolder for schema : Directory /var/www/html/folder-name/development-www/schemas/newfolder already exists
# ls -laR; cat newfolder/.skeema
.:
total 12
drwxr-xr-x.  3 root root 4096 Aug  6 21:17 .
drwxr-xr-x. 24 root root 4096 Aug  6 21:17 ..
drwxr-xr-x.  2 root root 4096 Aug  6 21:17 newfolder

./newfolder:
total 12
drwxr-xr-x. 2 root root 4096 Aug  6 21:17 .
drwxr-xr-x. 3 root root 4096 Aug  6 21:17 ..
-rw-r--r--. 1 root root   88 Aug  6 21:17 .skeema
[production]
host=localhost
socket=/var/lib/mysql/mysql.sock
user=root
flavor=mysql:8.0```

I believe the version is the latest one

```# /root/go/bin/skeema --version
skeema version 1.0.3-dev (snapshot build from source)```

```# uname -a
Linux hidden-hostname 3.10.0-693.17.1.el7.x86_64 #1 SMP Thu Jan 25 20:13:58 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```
I do have multiple databases, and If I add the --schema database_name option I don't get the error anymore
```# mkdir schemas; cd schemas; /root/go/bin/skeema init production --debug --dir newfolder --schema database_name -S /var/lib/mysql/mysql.sock -h localhost -u root -p
Enter password:
2018-08-06 21:38:01 [INFO]  Creating and using host dir /var/www/html/lumminary.com/development-www/schemas/newfolder for localhost:/var/lib/mysql/mysql.sock; skipping schema-level subdirs

2018-08-06 21:38:01 [INFO]  Populating /var/www/html/folder-name/development-www/schemas/newfolder

2018-08-06 21:38:01 [DEBUG] Exit code 0 (SUCCESS)
```

I do know you don't have support for MySQL 8.0 but I still wanted to give it a try, hoping it will work.
It would be amazing if you can make it work with MySQL 8 as I love the way you architect the entire solution.

Please let me know if I can help with anything else

**Comment by evanelias** [comment]
Very helpful, thanks! It seems that Skeema is failing to query the schema name correctly; it's getting a blank string. So then when it tries to build a path for the schema's subdir, it appends the schema name (blank) to the end of the host dir, and ends up just getting the host dir again... which fails since the dir already exists.

Looking through the MySQL 8.0 manual, I believe I found the culprit. Skeema queries `information_schema` extensively, and 8.0 changed the casing on column names in queries on those views. e.g. instead of getting back `schema_name`, 8.0 by default will return `SCHEMA_NAME`, which wasn't being handled by Skeema properly. (in [mysql 8.0 changes doc](https://dev.mysql.com/doc/refman/8.0/en/data-dictionary-usage-differences.html), search the page for "result set headers")

I've put a potential fix in master -- easiest way to upgrade is `go get -u github.com/skeema/skeema`. Please give this a try and let me know if it fixes this problem -- thanks! I haven't actually tested against 8.0 yet or made any other 8.0 compatibility improvements though, so it's possible other things will break. I'll aim for having full 8.0 support in Skeema 1.1.x, hopefully within the next month or so.

**Comment by evanelias** [comment]
Glad to hear that fixed the initial problem! Bummer that there are other 8.0 incompatibilities lurking in there, but not surprising. I'll close this issue out since the initial problem is fixed, but will comment on here once 8.0 is fully supported.

Regarding your other question, yes exactly, you can just edit the .skeema file in the schema-level directory. By default, skeema puts schema names outside of any named environment section, which means they apply to all environments (unless overridden by a specific environment). But you can move them into an environment section manually. For example, you could make it look like this:
```
[production]
schema=main_production

[development]
schema=main_development
```

That way, the appropriate name would be used depending on which environment you specified on the CLI for any given command.

**Comment by evanelias** [comment]
@FlaviuRadulescu I've just released [Skeema v1.0.4](https://github.com/skeema/skeema/releases/tag/v1.0.4), which fixes all known MySQL 8.0 incompatibilities exposed by Skeema's integration test suite.

There's possibly still some other edge case problems lurking in there, but it should be safe to try and should support common use-cases. That said, please feel free to open a new issue if you encounter anything odd in 8.0. Thanks again for reporting the issue!

**Comment by FlaviuRadulescu** [comment]
Dear Evan,

This is great news, I will test in the following days, and I will let you know if I encounter any problem

Thank you for your great effort.


---
FLAVIU RADULESCU | Lumminary  |  lumminary.com<https://lumminary.com/>

PLEASE NOTE:  This email and any file transmitted are confidential and/or legally privileged and intended only for the person(s) directly addressed. If you are not the intended recipient, any use, copying, transmission, distribution, or other forms of dissemination is strictly prohibited. If you have received this email in error, please notify the sender immediately and permanently delete the email and files, if any.
